Refactor cert utils into one pkg, add funcs from bootkube for kubeadm to use

errordeveloper opened the pull request:
**What this PR does / why we need it**:

We have ended-up with rather incomplete and fragmented collection of utils for handling certificates. It may be worse to consider using `cfssl` for doing all of these things, but for now there is some functionality that we need in `kubeadm` that we can borrow from bootkube. It makes sense to move the utils from bookube into core, as discussed in #31221.

**Special notes for your reviewer**: I've taken the opportunity to review names of existing funcs and tried to make some improvements in that area (with help from @peterbourgon).

**Release note**:

``` release-note
NONE
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32876)

<!-- Reviewable:end -->

errordeveloper committed:
Refactor utils that deal with certs

- merge `pkg/util/{crypto,certificates}`
- add funcs from `github.com/kubernetes-incubator/bootkube/pkg/tlsutil`
- ensure naming of funcs is fairly consistent

errordeveloper commented:
cc @gtank @peterbourgon @yifan-gu @mikedanese

yifan-gu commented on pkg/util/cert/cert.go:
```diff
@@ -0,0 +1,282 @@
+	return pem.EncodeToMemory(&block)
+}
+
+func NewSelfSignedCACertificate(cfg CertConfig, key *rsa.PrivateKey) (*x509.Certificate, error) {
```
Nit: To be consistent, maybe change to `MakeSelfSignedCACertificate`? Or the other way around, change `MakeXXX` to `NewXXX`?

errordeveloper replied:
New is for funcs that return pointer to struct, and make is for ones that return a copy of the struct, isn't it?

peterbourgon replied:
Yep! Idiomatically in Go, NewFoo constructors return a reference type, and MakeFoo a value type.

luxas replied:
I think this is okay, as it returns a reference type

yifan-gu commented on pkg/util/cert/cert.go:
```diff
@@ -0,0 +1,282 @@
+
+// NewFromPEM returns the x509.Certificates contained in the given PEM-encoded byte array
+// Returns an error if a certificate could not be parsed, or if the data does not contain any certificates
+func NewFromPEM(pemCerts []byte) ([]*x509.Certificate, error) {
```
Nit: This name itself doesn't clearly reflects that it's making a Cert?
Maybe `NewCertsFromPEM` or `MakeCertsFromPEM`.

And btw we can move this above to group with `ParsePEMEncodedCACert` and `ParsePEMEncodedPrivateKey` ?

peterbourgon replied:
It's in package cert and will be called as `cert.NewFromPEM` — does cert not imply an x509.Certificate?

errordeveloper replied:
I've realised that this belongs to `pem.go` and it's called `ParseCertsPEM` now, to emphasise that it returns multiple certs.

errordeveloper replied:
It's true that this package is all about certs, but actually it also deals with keys. Externally it would look more like `certutil.PastseCertsPEM`, which seems okay to me and matches other utils, like `nodeutil`.

luxas replied:
Okay, @errordeveloper, can you change this function to `NewCertsFromPEM` as @yifan-gu suggested or move it to `pem.go` and call it `ParseCertsPEM`?

errordeveloper replied:
That's what it became in the latest commit.

yifan-gu commented:
LGTM except small nits.

errordeveloper commented:
@yifan-gu I'll have another go at making it a bit more consistent. It's still a mixed bag of funcs, but I think some more improvements can be done to it as is.

errordeveloper commented:
I've rebased and made a few more improvements. Some of the funcs from bootkube where not quite useful, as we already had something similar and kubeadm won't be using those. I've also moved all parsing and encoding funcs into `pem.go` and simplified naming of those funcs.

luxas commented:
`verify-golint` failed. Please fix that.

Also maybe squash, if you don't think your commit grouping does matter a lot...

luxas commented on pkg/util/cert/cert.go:
```diff
@@ -0,0 +1,186 @@
+// The certificate will be created with file mode 0644. The key will be created with file mode 0600.
+// If the certificate or key files already exist, they will be overwritten.
+// Any parent directories of the certPath or keyPath will be created as needed with file mode 0755.
+func GenerateSelfSignedCert(host, certPath, keyPath string, alternateIPs []net.IP, alternateDNS []string) error {
```
I guess it would be more consistent to take a `CertConfig` as a param instead of alternateIPs and alternateDNS.
=> `GenerateSelfSignedCert(cfg CertConfig, certPath, keyPath string)`

errordeveloper replied:
Sure, but this is an old function and changing it would imply more changes outside of this package.

luxas reviewed (commented):
I think this is a good start. Surely we can do even better in the future regarding `cfssl` and even more refactoring/naming, but this is a really good start.

Also, given that this is required for `kubeadm` (which we'll try to merge in a few days, we should get this merged into master as soon as possible.

errordeveloper commented:
I've squashed and made linter happy.

philips commented:
cc @aaronlevy

aaronlevy commented on pkg/util/cert/cert.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// NewSignedCert creates a signed certificate using the given CA certificate and key
+func NewSignedCert(cfg Config, key *rsa.PrivateKey, caCert *x509.Certificate, caKey *rsa.PrivateKey) (*x509.Certificate, error) {
```
nit: This doesn't seem to be used anywhere.

aaronlevy commented on pkg/util/cert/cert.go:
```diff
@@ -0,0 +1,190 @@
+		NotBefore:    caCert.NotBefore,
+		NotAfter:     time.Now().Add(duration365d).UTC(),
+		KeyUsage:     x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
```
nit: Doesn't look like this is used, but I know during the tls bootstrap discussion, others more versed in security didn't like having both server and client key usage in a single cert.

errordeveloper replied:
Happy to write to funcs for server and client, but not sure what's best for kubelet, I think it needs both. (Or does it?)

errordeveloper replied:
I'll add `ServerUsage bool, ClientUsage bool` to `Config` struct to control this, if that'd be okay?

aaronlevy replied:
I called it out as a nit because I'm relaying a concern that was expressed during the kubelet tls bootstrap discussion, but now digging back through https://github.com/kubernetes/kubernetes/pull/30094 I'm actually unsure of the final state. However, it seemed @gtank had no personal issue with issuing certs with both usages (and I trust his judgement). So I wouldn't block on changes here - but I'm not a pki infra expert either.

aaronlevy replied:
As a side note -- if you're not adding `--kubelet-client-certificate` `--kubelet-client-key` to the api-server -- the server usage isn't necessary (api-server will not validate the server cert in the kubelet api anyway).

aaronlevy replied:
Sorry to keep commenting, but I'm actually curious how this will be used in kubeadm in combination with the tls bootstrap? The CSR dance should get you at minimum a cert with client usage -- is this just meant for the kubelet api?

aaronlevy commented on pkg/util/cert/cert.go:
```diff
@@ -0,0 +1,190 @@
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA: true,
```
Why does this need to be a CA?

errordeveloper replied:
I didn't write this, just copied from bootkube. Happy to remove, thanks for pointing this out.

aaronlevy replied:
There is no `GenerateSelfSignedCert` in bootkube -- but does look like this is from exiting kubernetes code: https://github.com/kubernetes/kubernetes/blob/0986a01f4f2a00dcb5d3e58832c5f269f9788c62/pkg/util/crypto/crypto.go#L65

Still not sure why it would need to be a CA, but this already existed.

errordeveloper replied:
Yes, thanks for double checking. This is one of the old ones, it's used by apiserver and kubelet. I am only vaguely familiar with what it gets the apiserver - it goes and creates a cert if you didn't specify one, but I am not sure of how much use this behaviour is to anyone really, to me it seems complete useless and has usability bugs like https://github.com/kubernetes/kubernetes/issues/19309 and https://github.com/kubernetes/kubernetes/issues/19310. There is also a comment about how badly uneducated the apiserver is about what it's external names are, as it's not able to do anything about the extra SANs. I'd vote to rip it out completely.

krousey replied:
**Warning, not an expert**

My guess is that since this is also being used to self-sign in the call to `x509.CreateCertificate` below, that `IsCA` was set to true. The docs for the function call out self-signing specifically if both the parameters are the same, so would you get a valid self-signed cert without this? Could you test that?

@mikedanese Do you know more about this than me?

krousey replied:
After talking with @mikedanese, this is used to generate a cert authority to sign the rest of the cluster's certs.

mikedanese replied:
Ya, at the very least the name of this func isn't good. If this is used to generate the root CA for the cluster, the name should reflect that. Otherwise, we should not be minting a CA if it doesn't need to be.

errordeveloper commented:
Unused funcs will be used by kubeadm. All the new code was copied from
bootkube.

On Tue, 20 Sep 2016, 17:48 Aaron Levy, <notifications@github.com> wrote:

> *@aaronlevy* commented on this pull request.
> ------------------------------
>
> In pkg/util/cert/cert.go
> <https://github.com/kubernetes/kubernetes/pull/32876#pullrequestreview-788528>
> :
>
> > +		NotBefore:             now.UTC(),
> +		NotAfter:              now.Add(duration365d * 10).UTC(),
> +		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
> +		BasicConstraintsValid: true,
> +		IsCA: true,
> +	}
> +
> +	certDERBytes, err := x509.CreateCertificate(cryptorand.Reader, &tmpl, &tmpl, key.Public(), key)
> +	if err != nil {
> +		return nil, err
> +	}
> +	return x509.ParseCertificate(certDERBytes)
> +}
> +
> +// NewSignedCert creates a signed certificate using the given CA certificate and key
> +func NewSignedCert(cfg Config, key *rsa.PrivateKey, caCert *x509.Certificate, caKey *rsa.PrivateKey) (*x509.Certificate, error) {
>
> nit: This doesn't seem to be used anywhere.
> ------------------------------
>
> In pkg/util/cert/cert.go
> <https://github.com/kubernetes/kubernetes/pull/32876#pullrequestreview-788528>
> :
>
> > +	if err != nil {
> +		return nil, err
> +	}
> +
> +	certTmpl := x509.Certificate{
> +		Subject: pkix.Name{
> +			CommonName:   cfg.CommonName,
> +			Organization: caCert.Subject.Organization,
> +		},
> +		DNSNames:     cfg.AltNames.DNSNames,
> +		IPAddresses:  cfg.AltNames.IPs,
> +		SerialNumber: serial,
> +		NotBefore:    caCert.NotBefore,
> +		NotAfter:     time.Now().Add(duration365d).UTC(),
> +		KeyUsage:     x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
> +		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
>
> nit: Doesn't look like this is used, but I know during the tls bootstrap
> discussion, others more versed in security didn't like having both server
> and client key usage in a single cert.
> ------------------------------
>
> In pkg/util/cert/cert.go
> <https://github.com/kubernetes/kubernetes/pull/32876#pullrequestreview-788528>
> :
>
> > +	if err != nil {
> +		return err
> +	}
> +
> +	template := x509.Certificate{
> +		SerialNumber: big.NewInt(1),
> +		Subject: pkix.Name{
> +			CommonName: fmt.Sprintf("%s@%d", host, time.Now().Unix()),
> +		},
> +		NotBefore: time.Now(),
> +		NotAfter:  time.Now().Add(time.Hour * 24 * 365),
> +
> +		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
> +		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
> +		BasicConstraintsValid: true,
> +		IsCA: true,
>
> Why does this need to be a CA?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32876#pullrequestreview-788528>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS9rgZOEwncBbk_bKkMJbsSBZwS7bks5qsA57gaJpZM4J_F-6>
> .
>

lavalamp commented:
Unused functions need tests that will fail if they get removed, and the test (or somewhere) needs a comment stating why the function is included.

luxas commented:
@lavalamp @krousey Are you okay with this change?
@errordeveloper surely can fix nits later; we just need this in order to get kubeadm in as well

krousey commented:
@luxas I just finished my first read through. I think there are 2 open concerns. @lavalamp pointed out a lack of tests, and this is particularly useful when you have functions that aren't used yet. @aaronlevy has an open question about why a cert is a CA, and I also want to know why.

errordeveloper commented:
Thise are pretty thin stdlib wrappers and the plan is to replace these with
cfssl, so I am not sure what sort of test are needed. Also, kubeadm will
merged shortly after, so we aren't going to have unused funcs hanging
around. I will adress the CA question in detail as soon as I get back to
keyboard.

On Tue, 20 Sep 2016, 20:05 krousey, <notifications@github.com> wrote:

> @luxas <https://github.com/luxas> I just finished my first read through.
> I think there are 2 open concerns. @lavalamp <https://github.com/lavalamp>
> pointed out a lack of tests, and this is particularly useful when you have
> functions that aren't used yet. @aaronlevy <https://github.com/aaronlevy>
> has an open question about why a cert is a CA, and I also want to know why.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32876#issuecomment-248400587>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWSwrDcS6rZ6g3LbV-oXz43AnnqEFBks5qsC6CgaJpZM4J_F-6>
> .
>

errordeveloper commented:
> Sorry to keep commenting, but I'm actually curious how this will be used in kubeadm in combination with the tls bootstrap? The CSR dance should get you at minimum a cert with client usage -- is this just meant for the kubelet api?

@aaronlevy only `kubeadm init` needs [this stuff](https://github.com/errordeveloper/kubernetes/blob/kubeadm/cmd/kubeadm/app/master/pki.go#L135), `kubeadm join` calls CSR API.

errordeveloper commented:
@krousey what's your judgement?

lukemarsden commented:
@aaronlevy could you take a look at this today, please?

lukemarsden commented:
@mikedanese friendly ping, you said you would take a look at this and assign it to someone?

luxas commented:
He already did, and assigned it to @krousey

krousey reviewed (commented):
Sorry if this seems delayed. Just starting my day over here on the west coast.

lukemarsden commented:
No problem @krousey thanks for reviewing this!
